File: talleres/taller06/Taller06.py
"""
@author Miguel Angel Correa Manrique - Pablo Buitrago Jaramillo
Taller06 Estructura datos y algorítmos
"""
import numpy as np

#Exercise 1
class ArrayList:

    def __init__(self):
        self.size = 0
        self.DEFAULT_CAPACITY = 4
        self.elements = np.empty(self.DEFAULT_CAPACITY, int)

    def incrementDEFAULT_CAPACITY(self):
        self.DEFAULT_CAPACITY = self.DEFAULT_CAPACITY*2

    def size(self):
        return self.size

    def getValue(self, index):
        try:
            return self.elements[index]
        except ValueError:
            print('Value not found')
    
    #The asymptotic complexity for this method is O(n), so using this algorithm on a text editor is not the most suitable option. The asymptotyc 
    # complexcity to add n characters is O(n^n)
    def add(self, x):
        self.size = self.size+1
        if self.DEFAULT_CAPACITY > self.size:
            self.elements[self.size-1] = x
        
        else:
            ArrayList.incrementDEFAULT_CAPACITY(self)
            elements2 = np.empty(self.DEFAULT_CAPACITY, int)
            
            for i in range(0, self.size):
                elements2[i] = self.elements[i]
            
            elements2[self.size-1] = x
            self.elements = elements2

    #The asymptotic complexity for this method is O(n^2), so using this algorithm on a text editor is not the most suitable option. The asymptotyc 
    # complexcity to add n characters is O(n^n)
    def addInIndex(self, index, x):
        self.size = self.size+1
        if self.DEFAULT_CAPACITY > self.size:
            elements2 = np.empty(self.DEFAULT_CAPACITY, int)
            
            for i in range(0, index):
                elements2[i] = self.elements[i]
            
            elements2[index] = x
            
            for i in range(index+1, self.size):
                elements2[i] = self.elements[i]
            
            self.elements = elements2
        
        else:
            ArrayList.incrementDEFAULT_CAPACITY(self)
            elements2 = np.empty(self.DEFAULT_CAPACITY, int)

            for i in range(0, index):
                elements2[i] = self.elements[i]

            elements2[index] = x

            for i in range(index+1, self.size):
                elements2[i] = self.elements[i]

            self.elements = elements2

    def deleteChar(self, index):
        self.size = self.size-1
        self.elements[self.size+1] = 0

        for i in range(index, self.size+1):
            self.elements[i] = self.elements[i+1]


    def toString(self):
        return self.elements

File: talleres/taller06/Taller06.cpp
#include <iostream>
#include <cstdlib>
#include <string>

using namespace std;

// #1
class ArrayList{	
	public:
		int DEFAULT_SIZE = 4;
		int size;
		int *arr;


		void Init(){
			size = 0;
			arr = new int[DEFAULT_SIZE];
		}

		//The asymptotic complexity for this method is O(n), so using this algorithm on a text editor is not the most suitable option. The asymptotyc 
   		// complexcity to add n characters is O(n^n)
		void add(int x){	
			size += 1;	
		
			if(size >= DEFAULT_SIZE){

				DEFAULT_SIZE = DEFAULT_SIZE*2;

				int *arr2; //Why doesn't work with arr2[DEFAULT_SIZE] (?)

				arr2 = new int[DEFAULT_SIZE];
				
				for(int i = 0; i < size; i++) arr2[i] = arr[i];

				arr2[size-1] = x;

				arr = arr2;
	
			} else arr[size-1] = x;

		}

		//The asymptotic complexity for this method is O(n^2), so using this algorithm on a text editor is not the most suitable option. The asymptotyc 
   		// complexcity to add n characters is O(n^n)
		void addInIndex(int x, int index){

			size += 1;

			if(size >= DEFAULT_SIZE){
				DEFAULT_SIZE = DEFAULT_SIZE*2;

				int *arr2;

				arr2 = new int[DEFAULT_SIZE];

				int i; 

				for(i = 0; i < index; i++) arr2[i] = arr[i];

				arr2[index] = x;

				for(i = index; i < size; i++) arr2[i+1] = arr[i];

				arr = arr2;	

			} else {
			   	//Since there's no new size, no second array needed, elements are moved one place and then index is added to the array.
				for(int i = size; i >= index; i--) arr[i] = arr[i-1];
			
				arr[index] = x;
			}
		}


		void del(int index){
			size -= 1;		
			
			arr[size+1] = 0;

			int temp;

			for(int i = index; i <= size;i ++)  arr[i] = arr[i+1];
				
		}


		int get(int i){

		return arr[i];

		}


		void print(){

			cout << "[ ";

			for(int i = 0; i < size; i++) cout << arr[i] << " ";
			

			cout << "]" << endl;	

		}



		//Optional #2 addReverse Algorithm
		void addReverse(int x){
			size += 1;


			if(size < DEFAULT_SIZE) arr[DEFAULT_SIZE-size] = x;

			else {

				DEFAULT_SIZE = DEFAULT_SIZE*2;

				int *arr2;

				arr2 = new int[DEFAULT_SIZE];
			
				arr2[DEFAULT_SIZE-size] = x;

				for(int i = 1; i < DEFAULT_SIZE/2; i++) arr2[DEFAULT_SIZE-size+i] = arr[i];

				arr = arr2;

			}

			
		}

		


		void printReversed(){

			cout << "[ ";

                        for(int i = DEFAULT_SIZE-size; i < DEFAULT_SIZE ; i++)  cout << arr[i] << " ";
                        

                        cout << "]" << endl;

		}

};



//Optional #2 Reader
void readArray(ArrayList myArray){
	int x;
	while(true){


		cout << "Put integer value: ";

		cin >> x;
	

		if(x==-1) break;

		myArray.addReverse(x);
	}

	myArray.printReversed();




}



class ArrayListString{
	public:
		int DEFAULT_SIZE = 4;
                int size;
                string *arr;


	void Init(){

		size = 0;
		arr = new string[DEFAULT_SIZE];


	}


       	void add(string x){
        	size += 1;

                if(size >= DEFAULT_SIZE){

                	 DEFAULT_SIZE = DEFAULT_SIZE*2;

                         string *arr2; 

                         arr2 = new string[DEFAULT_SIZE];

                         for(int i = 0; i < size; i++) arr2[i] = arr[i];

                         	arr2[size-1] = x;

                                arr = arr2;

                        } else arr[size-1] = x;

                }


	void print(){
	
                        cout << "[ ";

                        for(int i = 0; i < size; i++) cout << arr[i] << "       ";

                        cout << "]" << endl;

	
	}
	



};




//Optional #3 Sequence
void sequence(int n){
	ArrayListString myArray;
	myArray.Init();

	string str;


	for(int i = 1; i <= n; i++){
		str = "";
		for(int j = 1; j<=i; j++){
				
			 str+= to_string(j) + ", ";

		}
		
	myArray.add(str);

	}

	myArray.print();	

}


int main(){


}
